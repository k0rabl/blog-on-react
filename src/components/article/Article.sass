.article
  margin-top: 25px
  margin-bottom: 25px
  .top
    margin-bottom: 15px
    display: flex
    justify-content: space-between
    .title
      font-size: 26px