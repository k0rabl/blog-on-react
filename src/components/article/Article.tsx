import React, { Component } from 'react'
import IArticle from '../../interfaces/IArticle'

class Article extends Component<IArticle> {
  constructor(props: IArticle){
    super(props)
  }

  render() {
    const { date, name, desc } = this.props
    
    return (
      <div className="container">
        <div className="top">
          <div className="title">{name}</div>
          <div className="date">{date.toDateString()}</div>
        </div>
        <div className="bottom">
          <div className="image"></div>
          <div className="descr">{desc}</div>
        </div>
      </div>
    )
  }
}

export default Article