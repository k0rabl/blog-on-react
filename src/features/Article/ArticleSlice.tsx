import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import Articles from '../../fixtures/Articles'
import IArticle from './component/IArticle'

export interface PaginationState {
  filteredArticles: IArticle[]
  articles: IArticle[]
}

const initialState: PaginationState = {
  filteredArticles: Articles,
  articles: Articles
}

export const articlesSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    handleRead: (state, action: PayloadAction<number>) => {
      state.articles =  state.articles.map((element: IArticle) => ({
        ...element,
        isRead: element.id === action.payload ? true : element.isRead
      }))
    },
    handleDeleteElement: (state, action: PayloadAction<number>) => {
      state.articles = state.articles.filter(element => {
        return element.id !== action.payload
      })
    },
    handleEditElement: (state, action: PayloadAction<IArticle>) => {
      state.articles = state.articles.map((element: IArticle) => {
        if(element.id === action.payload.id)  
          return action.payload

        return element
      })
    },
    handleAddElement: (state, action: PayloadAction<IArticle>) => {
      state.articles.push(action.payload)
    }
  }
})

export const { 
  handleRead, 
  handleDeleteElement,
  handleEditElement,
  handleAddElement
} = articlesSlice.actions
export default articlesSlice.reducer