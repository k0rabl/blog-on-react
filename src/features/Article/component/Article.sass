
.article
  margin-top: 25px
  margin-bottom: 25px
  width: 100%
  display: flex
  justify-content: space-between
  .leftImage
    width: 30%
    border: 1px solid #b10303
    margin-right: 15px
    display: flex
    justify-content: center
    align-items: center
    img
      width: 100%
      height: 200px
  .rightDescr
    width: 70%
  &Link
    color: inherit
  
  .top
    margin-bottom: 15px
    display: flex
    justify-content: space-between
    &-left
      display: flex
      align-items: center
      .title
        margin-right: 15px
        font-size: 26px
        color: #b10303
        transition: color .2s
        &.edit
          color: #B94A3E
          white-space: nowrap
          text-overflow: ellipsis
          max-width: 350px
          overflow: hidden
        &.read
          color: #B94A3E
        &:hover
          color: #742c24
      .btn
        &-delete
          margin-right: 15px