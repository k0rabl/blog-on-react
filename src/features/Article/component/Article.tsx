import React, { Component } from 'react'
import IArticle from './IArticle'

import './Article.sass'
import { Link } from 'react-router-dom'

class Article extends Component<IArticle> {

  render() {
    const { id, date, name, desc, isRead } = this.props
    
    return (
      <Link to={`/${id}`} className="link articleLink">
        <div className="container article">
          <div className="top">
            <div className={`title ${isRead && 'read'}`}>{name}</div>
            <div className="date">{date}</div>
          </div>
          <div className="bottom">
            <div className="image"></div>
            <div className="descr">{desc}</div>
          </div>
        </div>
      </Link>
    )
  }
}

export default Article