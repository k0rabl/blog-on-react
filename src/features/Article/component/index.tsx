import React, { PureComponent } from 'react'
import { ArticleProps, IState } from './IArticle'

import './Article.sass'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { RootState } from '../../../redux/store'
import { handleDeleteElement } from '../ArticleSlice'
import { Modal } from '../../Modal'
import AlertContext from '../../../AlertContext'

class Article extends PureComponent<ArticleProps, IState> {
  static contextType = AlertContext

  constructor(props: ArticleProps){
    super(props)

    this.state = {
      opened: false
    }
  }

  handleDelete = () => {
    const { setOpen, setType } = this.context 
    const { id, handleDeleteElement } = this.props
    
    handleDeleteElement(Number(id))
    
    setOpen(true)
    setType('Delete')
    this.setState({opened: false})
  }



  handleShowModal = () => {
    const { opened } = this.state
    this.setState({opened: !opened})
  }

  render() {
    const { 
      id, 
      date, 
      name, 
      desc, 
      isRead,
      editMode
    } = this.props   
    
    return (
      <div className="container article">
        <div className="top">
          <div className="top-left">
            
            <Link to={`/${id}`} className="link articleLink">
              <div className={`title ${isRead && 'read'} ${editMode && 'edit'}`}>{name}</div>
            </Link>
            
            {editMode && 
              <div className="buttons">
                <button  className="waves-effect waves-light redC-bg lighten-3 btn btn-delete" onClick={this.handleShowModal}>
                  <i className="material-icons center">delete</i>
                </button > 
                
                {/* <button  className="waves-effect waves-light redC-bg lighten-3 btn btn-edit" onClick={() => console.log('edit')}>
                  <i className="material-icons center">edit</i>
                </button > */} 
              </div>
            }
          </div>
          
          <div className="date">{date}</div>
        </div>
        <div className="bottom">
          <div className="image"></div>
          <div className="descr">{desc}</div>
        </div>
        <Modal 
          opened={this.state.opened}
          onClose={this.handleShowModal} 
          onAccept={this.handleDelete} 
          type='delete'
        />
      </div>
    )
  }
}


const mapStateToProps = (state: RootState) => ({
  editMode: state.edit.editMode,
  articles: state.search.filteredArticles
})

const mapDispatch = { handleDeleteElement }

export default connect(mapStateToProps, mapDispatch)(Article)