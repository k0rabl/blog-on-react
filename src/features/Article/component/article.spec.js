import React from "react"
import Article from  './index'
import { mount } from 'enzyme';

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

import {
  BrowserRouter as Router,
} from "react-router-dom"


describe('<Article />', () => {
  const initialState = { 
    edit: {        
      editMode: true,
    }
  };
  const mockStore = configureStore();
  let store, component;
  
  beforeEach(() => {
    store = mockStore(initialState)

    component = mount(
      <Provider store={store}>
        <Router>
          <Article 
            match={{params: {id: 1}}}
            />
        </Router>
      </Provider>
    )
  })

  it('+++ render the component', () => {
    expect(component.find(Article).length).toEqual(1)
  })

  
  it('+++ check Button delete', () => {
    component
      .find('button.btn-delete')
      .simulate('click');
  })
})