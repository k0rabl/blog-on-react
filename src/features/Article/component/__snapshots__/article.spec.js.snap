// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> +++ render the component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ModalProvider>
    <BrowserRouter>
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(Article)
          match={
            Object {
              "params": Object {
                "id": 1,
              },
            }
          }
        >
          <Article
            editMode={true}
            handleDeleteElement={[Function]}
            match={
              Object {
                "params": Object {
                  "id": 1,
                },
              }
            }
          >
            <div
              className="container article"
            >
              <div
                className="leftImage"
              >
                <span>
                  No image
                </span>
              </div>
              <div
                className="rightDescr"
              >
                <div
                  className="top"
                >
                  <div
                    className="top-left"
                  >
                    <Link
                      className="link articleLink"
                      to="/undefined"
                    >
                      <LinkAnchor
                        className="link articleLink"
                        href="/undefined"
                        navigate={[Function]}
                      >
                        <a
                          className="link articleLink"
                          href="/undefined"
                          onClick={[Function]}
                        >
                          <div
                            className="title undefined edit"
                          />
                        </a>
                      </LinkAnchor>
                    </Link>
                    <div
                      className="buttons"
                    >
                      <button
                        className="waves-effect waves-light redC-bg lighten-3 btn btn-delete"
                        onClick={[Function]}
                      >
                        <i
                          className="material-icons center"
                        >
                          delete
                        </i>
                      </button>
                    </div>
                  </div>
                  <div
                    className="date"
                  />
                </div>
                <div
                  className="bottom"
                >
                  <div
                    className="descr"
                  />
                </div>
              </div>
            </div>
          </Article>
        </Connect(Article)>
      </Router>
    </BrowserRouter>
  </ModalProvider>
</Provider>
`;
